File: day03/part2.nim
import nre, options, strutils, sequtils, parseutils

var grid = newSeqWith(1000, newSeq[int](1000))

proc is_unique(left: int, top: int, width: int, height: int): bool = 
  for row in top .. (top + height - 1):
    for col in left .. (left + width - 1):
      if grid[row][col] > 1:
        return false
  return true

var total = 0
var claims = newSeqWith(1000, newSeq[int](5))
for line in "input".lines():
  let f = line.find(re"""#(\d+) \@ (\d+),(\d+): (\d+)x(\d+)""")
  var claim = map(f.get.captures.toSeq(), proc(x: string): int = parseInt(x))
  var id = claim[0]
  var left = claim[1]
  var top = claim[2]
  var width = claim[3]
  var height = claim[4]
  claims.add(@[id, left, top, width, height])
  
  for row in top .. (top + height - 1):
    for col in left .. (left + width - 1):
      grid[row][col] += 1

for claim in claims:
  var uniq = is_unique(claim[1], claim[2], claim[3], claim[4])
  if uniq:
    echo claim[0]

File: day03/part2.py
import re


class Grid(object):
    def __init__(self, size=1000):
        self.grid = [[0 for i in range(size)] for u in range(size)]

    def add(self, id, left, top, width, height):
        for row in range(top, top+height):
            for col in range(left, left+width):
                self.grid[row][col] += 1

    def uniq(self, id, left, top, width, height):
        t = None
        for row in range(top, top+height):
            for col in range(left, left+width):
                if self.grid[row][col] > 1:
                    return False
        return True


grid = Grid()
claims = []
r = re.compile(r"""#(\d+) \@ (\d+),(\d+): (\d+)x(\d+)""")
for line in open("input", "r").readlines():
    f = r.findall(line.strip())
    claim = list(map(int, f[0]))
    id, left, top, width, height = claim
    claims.append(claim)
    grid.add(id, left, top, width, height)
for claim in claims:
    if grid.uniq(*claim):
        print(claim[0])
        break
